posthog/clickhouse/log_entries.py

```diff
@@ -0,0 +1,68 @@
+from posthog.clickhouse.kafka_engine import KAFKA_COLUMNS, kafka_engine, ttl_period
+from posthog.clickhouse.table_engines import ReplacingMergeTree
+from posthog.kafka_client.topics import KAFKA_LOG_ENTRIES
+from posthog.settings import CLICKHOUSE_CLUSTER, CLICKHOUSE_DATABASE
+
+LOG_ENTRIES_TABLE = "log_entries"
+LOG_ENTRIES_TTL_WEEKS = 1
+
+LOG_ENTRIES_TABLE_BASE_SQL = """
+CREATE TABLE IF NOT EXISTS {table_name} ON CLUSTER '{cluster}'
+(
+    team_id UInt64,
+    log_source LowCardinality(String),
+    log_source_id VARCHAR,
+    instance_id VARCHAR,
+    timestamp DateTime64(6, 'UTC'),
+    level LowCardinality(String),
+    message VARCHAR
+    {extra_fields}
+) ENGINE = {engine}
+"""
+
+LOG_ENTRIES_TABLE_ENGINE = lambda: ReplacingMergeTree(LOG_ENTRIES_TABLE, ver="_timestamp")
+LOG_ENTRIES_TABLE_SQL = lambda: (
+    LOG_ENTRIES_TABLE_BASE_SQL
+    + """ORDER BY (team_id, log_source, log_source_id, instance_id, timestamp)
```

Comment:
I've been told its best to have the order by in increasing order of cardinality in which case I guess this would be: 

```suggestion
    + """ORDER BY (team_id, log_source, timestamp, log_source_id, instance_id)
```

(maybe)

Reply:
I would expect there to be more timestamps than `log_source_id` 🤔 At least in the batch exports use case a single one can generate multiple timestamps.

Reply:
Your order by is super close as it is. You really want a blend of cardinality + how used it will be in the predicates of the queries hitting the table.

I would definitely add a partition if you can:
-  `toStartOfHour(timestamp)` I imagine we will always filter by some time or sort by time? 24*90 is constant 2160 partitions which is reasonable.
This also makes TTL much cheaper if we TTL by hour since we won't be mutating but dropping partitions.

The way I think about is first lay out how you will query the table. What will you almost always be filter on?
- `team_id` ✅ 
- `log_source` ✅ 
- `log source_id` ? Maybe I'm not sure exactly how this relates to log_source, should be good to go though
- `instance_id` ✅ 
- `timestamp` ✅  This is only really good here if use the `toStartOfHour(timestamp)` partition above

Reply:
Ooh good call on the partition!

-----

If I'm understanding correctly then `log_source_id`  maps to `session_id` for replay. And we'd never provide an instance id 

But I'm going to be selecting that not querying it

So, I think the only way I can help improve this schema is when we're debugging something in a few months time and learn something new and realise what we should have done instead 🤣

Reply:
For batch exports, the are two query patterns:
* `WHERE team_id = <team_id> AND log_source = "batch_exports" AND log_source_id = <batch_export_id> AND instance_id = <run_id> ORDER BY timestamp`
* `WHERE team_id = <team_id> AND log_source = "batch_exports" AND log_source_id = <batch_export_id> ORDER BY timestamp`

The only difference between the two is whether we include or not `instance_id`: When showing the user all logs in the UI we will exclude it, but we may want to add a filter for logs of specific runs.

Reply:
Agree on adding a partition anyways 👍 I can add that in my next commit

Reply:
Forgot to add that we do allow users to filter on log level, so the two query patterns I described above could include  a `AND level = <level>`.

Reply:
sgtm! I think this is good enough for now. I don't think we _really_ need to include the `level` as a key. The more keys we have the slower the inserts and merges.